### PR TITLE
[docs] Add Mixed Precision detailed docs

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Re-introduces a mixed precision set of docs. This is primarily to explain bfloat16, and give an overview of the various options. There are some duplication of docs however I think this is fine primarily for discoverability.

We also need a reference point for bfloat16 to share :)

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/advanced/mixed_precision.rst [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+Mixed Precision Training
+========================
+
+Mixed precision combines the use of both FP32 and lower bit floating points (such as FP16) to reduce memory footprint during model training, resulting in improved performance.
+
+Lightning offers mixed precision training for GPUs, CPUs and TPUs.
+
+.. note::
+
+    In some cases it is important to remain in FP32 for numerical stability, so keep this in mind when using mixed precision.
```
Comment (justusschock) [inline docs/source/advanced/mixed_precision.rst:17]: I thought that's what mixed precision does? Only converting ops that are safe to do in fp16?

Or maybe I understood the note wrong? :D

File: docs/source/guides/speed.rst [hunk 2]
```diff
@@ -210,7 +210,7 @@ Mixed precision (16-bit) training
 
 Mixed precision combines the use of both 32 and 16 bit floating points to reduce memory footprint during model training, resulting in improved performance, achieving +3X speedups on modern GPUs.
 
-Lightning offers mixed precision or 16-bit training for GPUs and TPUs.
+Lightning offers mixed precision training for GPUs, CPUs and TPUs.
```
Comment (justusschock) [inline docs/source/guides/speed.rst:213]:
```suggestion
Lightning offers mixed precision training for CPUs and GPUs as well as 16-bit training for TPUs.
```

TPU isn't amp, but real half-prec I think, right?

**Review by tchaton — APPROVED** [review]
LGMT !

File: docs/source/advanced/mixed_precision.rst (hunk 1 above)
Comment (SeanNaren) [reply docs/source/advanced/mixed_precision.rst:17]: i know for torchpoint32 you have to remain in FP32 when doing scatter ops during computation. But that is as far as my knowledge goes :D

File: docs/source/guides/speed.rst (hunk 2 above)
Comment (SeanNaren) [reply docs/source/guides/speed.rst:213]: it's bfloat16, so will update

File: docs/source/advanced/mixed_precision.rst (hunk 1 above)
Comment (awaelchli) [reply docs/source/advanced/mixed_precision.rst:17]: I think the note is just trying to answer the question why we can't do true FP16 training in many cases.

File: docs/source/advanced/mixed_precision.rst [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+
+.. _amp:
+
+Mixed Precision Training
+========================
+
+Mixed precision combines the use of both FP32 and lower bit floating points (such as FP16) to reduce memory footprint during model training, resulting in improved performance.
+
+Lightning offers mixed precision training for GPUs and CPUs, as well as bfloat16 mixed precision training for TPUs.
```
Comment (awaelchli) [inline docs/source/advanced/mixed_precision.rst:13]: will we update this message only when 1.10 comes out? it's too early now?

File: docs/source/advanced/mixed_precision.rst [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+BFloat16 Mixed precision is similar to FP16 mixed precision, however we maintain more of the "dynamic range" that FP32 has to offer. This means we are able to improve numerical stability, compared to FP16 mixed precision. For more information see `this TPU performance blog post <https://cloud.google.com/blog/products/ai-machine-learning/bfloat16-the-secret-to-high-performance-on-cloud-tpus>`__.
+
+Since BFloat16 is more stable than FP16 during training, we do not need to worry about any gradient scaling or nan gradient values that comes with using FP16 mixed precision.
+
+.. testcode::
+    :skipif: not _TORCH_BFLOAT_AVAILABLE
+
+    Trainer(gpus=1, precision="bf16")
+
+It is also possible to use BFloat16 Mixed Precision on the CPU, relying on MKLDNN under the hood.
```
Comment (awaelchli) [inline docs/source/advanced/mixed_precision.rst:53]:
```suggestion
It is also possible to use BFloat16 mixed precision on the CPU, relying on MKLDNN under the hood.
```
above in the text it was not capitalized

File: docs/source/advanced/mixed_precision.rst [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    :skipif: not _TORCH_CPU_AMP_AVAILABLE
+
+    Trainer(precision="bf16")
+
+NVIDIA APEX Mixed Precision
+---------------------------
+
+.. warning::
+
+    NVIDIA APEX has been deprecated in favour of native mixed precision. It is suggested to use the above native mixed precision rather than APEX unless you know what you're doing.
```
Comment (awaelchli) [inline docs/source/advanced/mixed_precision.rst:65]: is this us making this statement or is this info coming from pytorch?

**Review by awaelchli — APPROVED** [review]

File: docs/source/advanced/mixed_precision.rst (hunk 5 above)
Comment (SeanNaren) [reply docs/source/advanced/mixed_precision.rst:67]: nvm I read the docs wrong: https://nvidia.github.io/apex/amp.html#for-users-of-the-old-amp-api

will change this to recommend users to use autocast!

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:
